src/net/connection/quality.rs

```diff
@@ -1,18 +1,93 @@
-// TODO: Congestion avoidance.
-// To prevent congestion avoidance we need ot decide whether the connection is good or bad.
-// If the network of the client is bad we do not flood the router with small packets.
-// When network conditions are `Good` we send 30 packets per-second, and when network conditions are `Bad` we drop to 10 packets per-second.
+use std::time::Duration;
+use std::sync::RwLockWriteGuard;
+
+use net::connection::VirtualConnection;
+
+use net::NetworkConfig;
+use packet::CongestionData;
 
 /// Represents the quality of an network.
 pub enum NetworkQuality {
     Good,
     Bad,
 }
 
+/// This type helps with calculating the round trip time from any packet.
+/// It is able to smooth out the network jitter if there is any.
+pub struct NetworkQualityMeasurer {
+    config: NetworkConfig,
+}
+
+impl NetworkQualityMeasurer
+{
+    pub fn new(config: NetworkConfig) -> NetworkQualityMeasurer
+    {
+        NetworkQualityMeasurer { config }
+    }
+
+    /// This will calculate the round trip time (rtt) from the given acknowledgement.
+    /// Where after it updates the rtt from the given connection.
+    pub fn update_connection_rtt(&self, connection: &mut RwLockWriteGuard<VirtualConnection>, ack_seq: u16)
+    {
+        let mut smoothed_rrt = 0.;
+        {
+            let mut congestion_data = connection.congestion_avoidance_buffer.get_mut(ack_seq);
+            smoothed_rrt = self.get_smoothed_rtt(congestion_data);
+        }
+
+        connection.rtt = smoothed_rrt;
+    }
+
+    /// This will get the smoothed round trip time (rtt) from the time we last heard from an packet.
+    fn get_smoothed_rtt(&self, congestion_avoidance_entry: Option<&mut CongestionData>) -> f32 {
+        match congestion_avoidance_entry {
+            Some(avoidance_data) => {
+                let elapsed_time = avoidance_data.sending_time.elapsed();
+
+                let rtt_time = self.as_milliseconds(elapsed_time);
+
+                self.smooth_out_rtt(rtt_time)
+            },
+            None => { 0. }
+        }
+    }
+
+    /// Converts an duration to milliseconds.
+    ///
+    /// `as_milliseconds` is not supported yet supported in rust stable.
+    /// See this stackoverflow post for more info: https://stackoverflow.com/questions/36816072/how-do-i-get-a-duration-as-a-number-of-milliseconds-in-rust
+    fn as_milliseconds(&self, duration: Duration) -> u64
+    {
+        let nanos = duration.subsec_nanos() as u64;
+        (1000 * 1000 * 1000 * duration.as_secs() + nanos) / (1000 * 1000)
```

Comment:
I feel like this could be made more concise, but not sure how, and not sure if it is worth it.

Reply:
I am also not sure how. Rust will at one-time support converting duration to milliseconds but for now, I had to copy some code from StackOverflow, see the link in the comments.